File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) { // dp on trees - aditya verma std approach
        int[] res = {Integer.MIN_VALUE};
        int temp = helper(root, res);
        return Math.max(temp, res[0]);
    }
    
    private int helper(TreeNode node, int[] res){ //returns height of the current node
        if(node == null) return -1;
        
        int left = 1+ helper(node.left, res); // height of left child + current-left edge
        int right = 1+ helper(node.right, res);
        
        int temp = Math.max(left, right);
        res[0] = Math.max(res[0], Math.max(left+right, temp));
        
        return temp;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        val ar = IntArray(1)
        // kotlin syntax issue?
        
        // approach 1 - working
        // val temp = helper(root, ar)
        // return Math.max(ar[0], temp)
        
        // approach 2 - working
        return Math.max(helper(root, ar), ar[0])
        
        // approach 3 - doesn't  work // because order matters
        // return Math.max(ar[0], helper(root, ar))
    }
    
    fun helper(root : TreeNode?, ar: IntArray): Int {
        if(root == null) return -1;

        val left = 1 + helper(root.left, ar)
        val right = 1 + helper(root.right, ar)
        
        val currHeight = Math.max(left, right)
        val currDiameter = left + right
        ar[0] = Math.max(ar[0], Math.max(currDiameter, currHeight))
        return currHeight
    }
}